src/platform/BUILD.gn

```diff
@@ -160,6 +160,10 @@ if (chip_device_platform != "none") {
       "CHIP_DEVICE_LAYER_NONE=1",
       "CHIP_DEVICE_LAYER_TARGET=NONE",
     ]
+
+    if (current_os == "android") {
+      defines += [ "EXTERNAL_KEYVALUESTOREMANAGERIMPL_HEADER=\"controller/java/AndroidKeyValueStoreManagerImpl.h\"" ]
```

Comment:
We've made a number of contradictory decisions here

- Put the KVS interfaces in device layer
- Omit the device layer on Android
- Require the KVS on Android

And more fundamentally, we are using this device layer as both a "top layer" (it includes logic that depends on CHIP's internals) and "bottom layer" (which CHIP's internals depends on to access platform facilities) which is a cyclic structure.

Reply:
I realize this does not provide any suggestion of a path forward.

IMO - It's wrong to use the KVS if device layer == none when the KVS is part of the device layer. We shouldn't do that which means we shoudn't land this patch as-is.

I'd suggest we move KVS interface out of there, and have the device layer merely provide an instance of it. Then, require a KVS object be passed in at initialization for any object that requires one (i.e., use dependency injection). Similar to how we do for SystemLayer etc with device layer "none". Then implement Android KVS as a discrete build target which the Android port instantiates directly (i.e., not via device layer) and passes to the objects that need it.

We could also just give up on supporting "no" device layer and add one for Android. I don't like this option as much, because I think we **should** allow applications to control the lifecycles of all their objects which device layer does not allow.

Reply:
I believe the difficulty lies in having KVS as a singleton. Without that, dependency injection would work just fine and I would not need this device layer bits.

@rgoliver - is it possible to split KVS base class declaration from the actual implementation and not have it as a singleton?
I believe we currently do some casts to impl to avoid virtual functions and this forces us to have both interface and implementation always available. Is this true? Can we refactory, e.g. make methods pure virtual?

Reply:
As a note: the reason I added KVS into android is to try to unify our storage story. We can also live with a separate abstraction (we have a 'storage delegate' right now), however I would rather have our storage accessor classes flat.

Reply:
Build time dependency injection is also possible and in fact we do it in several places already.

Reply:
That said, I think it is wrong that we can't have more than one KVS on Android. Android is not a severly constrained platform, and we need to avoid overly constraining platforms where we have more resources available. These tightly constrained styles are **worse** on those platforms.

Reply:
I think that one thing we should keep in mind is that using virtual methods makes it more difficult for the linker to optimize out unused code as it must fill a full vtable even if some methods are never called. Not sure if LTO makes it any better.
Perhaps we could have an Android platform and still allow for some flexibility - the dependency injection might happen at the platform layer and the KVS implementation for Android might just work as a proxy. This would keep the common CHIP code simple and added the complexity only for platforms which need it unless we do think that different CHIP components may need different KVS objects.

Reply:
I agree we should be careful not to give up ground on the constrained devices because there is none to give. At the same time, I'm worried about the impact the limitations we have there have on testability. Without runtime polymorphism, every variation requires an additional **build** instead of just an additional **test**. I think it's possible to solve it, though.

Reply:
I created https://github.com/project-chip/connectedhomeip/issues/5868 to track these changes.